lib/Service/MailManager.php

```diff
@@ -819,6 +855,50 @@ public function updateTag(int $id, string $displayName, string $color, string $u
 		return $this->tagMapper->update($tag);
 	}
 
+	public function deleteTag(int $id, string $userId, array $accounts) :Tag {
+		try {
+			$tag = $this->tagMapper->getTagForUser($id, $userId);
+		} catch (DoesNotExistException $e) {
+			throw new ClientException('Tag not found', 0, $e);
+		}
+
+		foreach ($accounts as $account) {
+			$this->deleteTagForAccount($id, $userId, $tag, $account);
+		}
+		return $this->tagMapper->delete($tag);
+	}
+
+	public function deleteTagForAccount(int $id, string $userId, Tag $tag, Account $account) :void {
+		try {
+			$messageTags = $this->messageTagsMapper->getMessagesByTag($id);
+			$messages = array_merge(... array_map(function ($messageTag) use ($account) {
+				return $this->getByMessageId($account, $messageTag->getImapMessageId());
+			}, [...$messageTags]));
+		} catch (DoesNotExistException $e) {
+			throw new ClientException('Messages not found', 0, $e);
+		}
+
+		$client = $this->imapClientFactory->getClient($account);
+
+		foreach ($messageTags as $messageTag) {
+			$this->messageTagsMapper->delete($messageTag);
+		}
+		$groupedMessages = [];
+		foreach ($messages as $message) {
+			$mailboxId = $message->getMailboxId();
+			if (array_key_exists($mailboxId, $groupedMessages)) {
+				$groupedMessages[$mailboxId][] = $message;
+			} else {
+				$groupedMessages[$mailboxId] = [$message];
+			}
+		}
+
+		foreach ($groupedMessages as $mailboxId => $messages) {
+			$mailbox = $this->getMailbox($userId, $mailboxId);
+			$this->tagMessageWithClient($client, $account, $mailbox, $messages, $tag, false);
+		}
+		$client->logout();
```

Comment:
use the client with a try-finally to ensure the connection is closed on error too